File: app/src/main/java/app/storytel/candidate/com/PostAdapter.java
package app.storytel.candidate.com;

import android.app.Activity;
import android.content.Intent;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.RequestManager;

import java.util.Random;

public class PostAdapter extends RecyclerView.Adapter<PostAdapter.PostViewHolder> {

    private PostAndImages mData;
    private RequestManager mRequestManager;
    private Activity mActivity;

    public PostAdapter(RequestManager requestManager, Activity activity) {
        mRequestManager = requestManager;
        mActivity = activity;
    }

    @Override
    public PostViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new PostViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.post_item, parent, false));
    }

    @Override
    public void onBindViewHolder(PostViewHolder holder, int position) {
        holder.title.setText(mData.getMPosts().get(position).getTitle());
        holder.body.setText(mData.getMPosts().get(position).getBody());
        int index = new Random().nextInt(mData.getMPhotos().size() - 1);
        String imageUrl = mData.getMPhotos().get(index).getThumbnailUrl();
        mRequestManager.load(imageUrl).into(holder.image);
        holder.body.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mActivity.startActivity(new Intent(mActivity, DetailsActivity.class));
            }
        });
    }

    public void setData(PostAndImages data) {
        mData = data;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return mData == null ? 0 : mData.getMPosts().size();
    }

    public class PostViewHolder extends RecyclerView.ViewHolder {
        TextView title;
        TextView body;
        ImageView image;

        public PostViewHolder(View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.title);
            body = itemView.findViewById(R.id.body);
            image = itemView.findViewById(R.id.image);
        }
    }
}

File: app/src/main/java/app/storytel/candidate/com/PostAdapter.kt
package app.storytel.candidate.com

import android.app.Activity
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import app.storytel.candidate.com.DetailsActivity
import app.storytel.candidate.com.PostAdapter.PostViewHolder
import com.bumptech.glide.RequestManager
import java.util.*

class PostAdapter(private val mRequestManager: RequestManager, private val mActivity: Activity) :
        RecyclerView.Adapter<PostViewHolder>() {

    private var mData: PostAndImages? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PostViewHolder {
        return PostViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.post_item, parent, false))
    }

    override fun onBindViewHolder(holder: PostViewHolder, position: Int) {
        holder.title.text = mData!!.mPosts.get(position).title
        holder.body.text = mData!!.mPosts.get(position).body
        val index = Random().nextInt(mData!!.mPhotos.size - 1)
        val imageUrl = mData!!.mPhotos[index].thumbnailUrl
        mRequestManager.load(imageUrl).into(holder.image)
        holder.body.setOnClickListener(
                View.OnClickListener { mActivity.startActivity(Intent(mActivity, DetailsActivity::class.java)) })
    }

    fun setData(data: PostAndImages?) {
        mData = data
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return if (mData == null) 0 else mData!!.mPosts.size
    }

    inner class PostViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var title: TextView
        var body: TextView
        var image: ImageView

        init {
            title = itemView.findViewById(R.id.title)
            body = itemView.findViewById(R.id.body)
            image = itemView.findViewById(R.id.image)
        }
    }
}
